### PR TITLE
MAINT: Change iter to iterable in array_fromiter

### LINKED ISSUE #10244: BUG: wrong or typo in signature for np.fromiter
gen = (0 for _ in range(3)
np.fromiter(iterable=gen,dtype='int') # fails
np.fromiter(iter=gen,dtype='int') # works

signature in docs and in core.multiarray declared as iterable

np.__version__
Out[8]: '1.13.3'

### PR DESCRIPTION
**Pull request by jrbourbeau**
Fixes #10244

Currently there's a mismatch between the docstring and function signature of `np.fromiter`. The docstring uses `np.fromiter(iterable, ...)` while the function signature uses `np.fromiter(iter, ...)`. Given that `iter` is a Python keyword, this PR changes `iter` to `iterable` in the function signature to match the existing docstring.

### CODE REVIEW

**Comment by eric-wieser** [comment]
> Given that iter is a Python keyword

**Comment by eric-wieser** [comment]
Arguments against merging:
* This is a breaking API change

Arguments for merging:
* This kwarg matches the name of the `itertools` arguments
* No one <sup>[citation needed]</sup> is going to call this by kwarg anyway, as they would only know `iter=` was supported by looking at the source - so the breaking change will not hit anyone

I think I'm +0.5 on merging this

**Comment by jrbourbeau** [comment]
> This is not true. iter is just a builtin

Good point, my mistake

**Comment by jrbourbeau** [comment]
Would you be more in favor of using `iter` over `iterable`?

**Comment by eric-wieser** [comment]
`iterable` is the better choice, were we adding a new function - the question is whether it's too late to change it in an existing one.

**Comment by mhvk** [comment]
A safer (but somewhat annoying) approach would be to check whether `iter` is in the keyword list, and, if so, raise a deprecation warning. I do think this can be a very short deprecation period.

**Comment by eric-wieser** [comment]
Which would look something like:

```
ret = PyArg_ParseTupleAndKeywords(the_old_kwargs);
if (ret == NULL) {
    PyErr_Clear();
    // Py_XDECREF all the args
    ret = PyArg_ParseTupleAndKeywords(the_new_kwargs);
}
else {
   if (DEPRECATE(the_message) < 0) {
       return NULL;
   }
}
```

**Comment by eric-wieser** [comment]
Maybe:
```
ret = PyArg_ParseTupleAndKeywords(the_new_kwargs);
if (ret == NULL) {
    PyErr_Fetch(&e_t, &e_v, &e_tb);
    // Py_XDECREF all the args
    ret = PyArg_ParseTupleAndKeywords(the_old_kwargs);
    if (ret == NULL) {
        PyErr_restore(e_t, e_v, e_tb);
        return NULL
    }
    if (DEPRECATE(the_message) < 0) {
       return NULL;
    }
}
```

**Comment by mhvk** [comment]
@eric-wieser - your second one is indeed better. I had thought of, before parsing, checking for `iter` in kwargs and, if present, give a deprecation message and put it back as `iterable`. But yours avoids changing the kwds or any slowdown for the regular case, so nicer.

**Comment by jrbourbeau** [comment]
@eric-wieser - Thanks for your feedback!

I was wondering about your `// Py_XDECREF all the args` comment. Since no new arguments are being added to `np.fromiter`, I just kept the same `Py_XDECREF(descr);` line. Not sure if I'm missing something here.

File: numpy/core/src/multiarray/multiarraymodule.c [hunk 1]
```diff
@@ -2192,18 +2192,31 @@ array_fromfile(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *keywds)
...
-        return NULL;
+
+        PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+        if (!PyArg_ParseTupleAndKeywords(args, keywds,
+                    "OO&|" NPY_SSIZE_T_PYFMT ":fromiter", kwlist_old,
+                    &iterable, PyArray_DescrConverter, &descr, &nin)) {
+            PyErr_Restore(ptype, pvalue, ptraceback);
+            return NULL;
+        }
+        if (DEPRECATE("The 'iter' keyword argument is deprecated. "
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/multiarraymodule.c]: Deprecations should have something like `/* NumPy 1.15, 2018-02-15 */` on the line before them - and preferably in the test too.

File: numpy/core/src/multiarray/multiarraymodule.c [hunk 2]
```diff
@@ -2192,18 +2192,31 @@ array_fromfile(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *keywds)
...
+                &iterable, PyArray_DescrConverter, &descr, &nin)) {
         Py_XDECREF(descr);
-        return NULL;
+
+        PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+        if (!PyArg_ParseTupleAndKeywords(args, keywds,
+                    "OO&|" NPY_SSIZE_T_PYFMT ":fromiter", kwlist_old,
+                    &iterable, PyArray_DescrConverter, &descr, &nin)) {
+            PyErr_Restore(ptype, pvalue, ptraceback);
+            return NULL;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/multiarraymodule.c]: I think you need to DECREF `descr` here too

File: numpy/core/src/multiarray/multiarraymodule.c [hunk 3]
```diff
@@ -2192,18 +2192,31 @@ array_fromfile(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *keywds)
...
+    PyObject *ptype, *pvalue, *ptraceback;
 
     if (!PyArg_ParseTupleAndKeywords(args, keywds,
                 "OO&|" NPY_SSIZE_T_PYFMT ":fromiter", kwlist,
-                &iter, PyArray_DescrConverter, &descr, &nin)) {
+                &iterable, PyArray_DescrConverter, &descr, &nin)) {
         Py_XDECREF(descr);
-        return NULL;
+
+        PyErr_Fetch(&ptype, &pvalue, &ptraceback);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/multiarraymodule.c]: A comment here saying "try the deprecated argument names" might be useful

**Comment by eric-wieser** [comment]
Hmm, maybe this doesn't work. What does:
```
np.fromiter(iter=range(10), dtype='this will totally blow up')
```
give?

**Comment by jrbourbeau** [comment]
It gives a `TypeError`

```python
>>> import numpy as np
>>> np.fromiter(iter=range(10), dtype='this will totally blow up')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Required argument 'iterable' (pos 1) not found
>>> np.fromiter(iter=range(10), dtype=int)
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
```

I'm not exactly sure what's causing this error though

**Comment by eric-wieser** [comment]
Yeah, that what I was worried would happen. I think @mhvk's approach is what we need after all.

**Comment by mattip** [comment]
@jrbourbeau will you be continuing with this? Do you agree that looking for "iter", if it is found emit a warning and use it as "iterable" is the correct approach?

**Comment by jrbourbeau** [comment]
@mattip I agree that checking for "iter" and raising a deprecation warning accordingly is a good way to go. Apologies for letting this PR linger, I've been busy with other work recently. @mattip you should feel free to open up a new PR if you're interesting in working on this issue.

**Comment by jrbourbeau** [comment]
This PR should now implement @mhvk approach of looking for an `"iter"` keyword, if found then emit a `DeprecationWarning` and use it as `"iterable"` instead. 

The example @eric-wieser previously asked about now gives a `TypeError` for the invalid `dtype`

```python
>>> import numpy as np
>>> np.fromiter(iter=range(10), dtype='this will totally blow up')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: data type "this will totally blow up" not understood
```

instead of an error for not providing an `iterable` positional argument.

**Comment by seberg** [comment]
While this is generally good (Thanks for the PR and sorry it never got merged!), I am going to close this.  I have to reasons for it:
1. It is unfortunately stale and requires some rebase (this even still has Python 2 support)
2. I actually think we should probably go one step further and make `iter` a position-only keyword argument now that Python has that.

If anyone disagrees please comment/reopen (or just open a new PR based on this and with the rebase/fixups done ;))

**Pull request closed without merging**